flink-connectors/flink-connector-rabbitmq/src/main/java/org/apache/flink/streaming/connectors/rabbitmq/RMQSource.java

```diff
@@ -268,54 +304,141 @@ public void run(SourceContext<OUT> ctx) throws Exception {
 		}
 	}
 
-	private class RMQCollector implements Collector<OUT> {
+	@Override
+	public void cancel() {
+		running = false;
+	}
+
+	@Override
+	protected void acknowledgeSessionIDs(List<Long> sessionIds) {
+		try {
+			for (long id : sessionIds) {
+				channel.basicAck(id, false);
+			}
+			channel.txCommit();
+		} catch (IOException e) {
+			throw new RuntimeException("Messages could not be acknowledged during checkpoint creation.", e);
+		}
+	}
+
+	@Override
+	public TypeInformation<OUT> getProducedType() {
+		return deliveryDeserializer.getProducedType();
+	}
 
+	/**
+	 * Special collector for RMQ messages.
+	 * Captures the correlation ID and delivery tag also does the filtering logic for weather a message has been
+	 * processed or not.
+	 */
+	private class RMQCollectorImpl implements RMQDeserializationSchema.RMQCollector<OUT> {
```

Comment:
What do you think about his implementation of the collector:

```
	private void processMessage(Delivery delivery, RMQCollectorImpl collector)
		throws IOException {
		AMQP.BasicProperties properties = delivery.getProperties();
		byte[] body = delivery.getBody();
		Envelope envelope = delivery.getEnvelope();
		collector.setFallBackIdentifiers(properties.getCorrelationId(), envelope.getDeliveryTag());
		deliveryDeserializer.deserialize(envelope, properties, body, collector);
	}


	/**
	 * Special collector for RMQ messages.
	 * Captures the correlation ID and delivery tag also does the filtering logic for weather a message has been
	 * processed or not.
	 */
	private class RMQCollectorImpl implements RMQDeserializationSchema.RMQCollector<OUT> {
		private final SourceContext<OUT> ctx;
		private boolean endOfStreamSignalled = false;
		private String correlationId;
		private long deliveryTag;
		private boolean customIdentifiersSet = false;

		private RMQCollectorImpl(SourceContext<OUT> ctx) {
			this.ctx = ctx;
		}

		@Override
		public void collect(OUT record) {
			if (!customIdentifiersSet) {
				boolean newMessage = setMessageIdentifiers(correlationId, deliveryTag);
				if (!newMessage) {
					return;
				}
			}

			if (isEndOfStream(record)) {
				this.endOfStreamSignalled = true;
				return;
			}
			ctx.collect(record);
		}

		public void setFallBackIdentifiers(String correlationId, long deliveryTag){
			this.correlationId = correlationId;
			this.deliveryTag = deliveryTag;
			this.customIdentifiersSet = false;
		}

		@Override
		public boolean setMessageIdentifiers(String correlationId, long deliveryTag){
			if (customIdentifiersSet) {
				throw new IllegalStateException("You can set only a single set of identifiers for a block of messages.");
			}

			this.customIdentifiersSet = true;
			if (!autoAck) {
				if (usesCorrelationId) {
					Preconditions.checkNotNull(correlationId, "RabbitMQ source was instantiated " +
						"with usesCorrelationId set to true yet we couldn't extract the correlation id from it !");
					if (!addId(correlationId)) {
						// we have already processed this message
						return false;
					}
				}
				sessionIds.add(deliveryTag);
			}
			return true;
		}

		boolean isEndOfStream(OUT record) {
			return endOfStreamSignalled || deliveryDeserializer.isEndOfStream(record);
		}

		public boolean isEndOfStreamSignalled() {
			return endOfStreamSignalled;
		}

		@Override
		public void close() {

		}
	}
```

Reply:
i think it's fine it's just missing the reset method to reset the `customIdentifiersSet, correlationId and deliveryTag` whenever  a new message is delivered.